<a href="https://colab.research.google.com/github/murpunk/Programming_2023/blob/main/29_05_23_Forest_9_lesson.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#0. Список литературы

1. Андреас Мюллер, Сара Гвидо. Введение в машинное обучение с помощью Python. Руководство для специалистов по работе с данными. С. 100—110.


2. Рашка Себастьян, Мирджалили Вахид Python и машинное обучение: машинное и глубокое обучение с использованием Python, scikit-learn и TensorFlow 2, 3-е изд.: Пер. с англ. СПб. : ООО "Диалектика", 2020. С. 135—138.

3. Плас Дж. Вандер Python для сложных задач: наука о данных и машинное обучение. СПб.: Питер, 2018. С. 481—487.

4. Жерон, Орельен. Прикладное машинное обучение с помощью Scikit-Learn и TensorFlow: концепции, инструменты и техники для создания интеллектуальных систем. Пер. с англ. СпБ.: ООО Альфа-книга: 2018. С. 239—250.

5. К.В. Воронцов Математические методы обучения по прецедентам (теория обучения машин). 141 с. (Voron-ML-1.pdf)

6. Машинное обучение (курс лекций, К.В.Воронцов): Многомерная линейная регрессия (курс лекций, К.В.Воронцов)
(machinelearning.ru)
 (лекция К.В. Воронцов “Линейные ансабли”:
https://www.youtube.com/watch?v=-wa43XNJfVI)


7. Крис Элбон Машинное обучение с использованием Python. Сборник рецептов. Пер. с англ. СПб. : БХВ-Петербург, 2019. 384 с.


#I. Прочитать литературу по теме

#II. Сделать задание из файла statement-forest.pdf.

### 1.  Загрузите данные из файла abalone.csv.


 Это датасет, в котором требуется предсказать возраст ракушки (число колец) по физическим измерениям.

In [1]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
# from sklearn import cross_validation # в задании нужно его использовать, но он не работает
from sklearn.model_selection import KFold

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
df = pd.read_csv("/content/drive/MyDrive/Программирование/abalone.csv")
df.head()

Sex  Length  Diameter  Height  WholeWeight  ShuckedWeight  VisceraWeight  \
0   M   0.455     0.365   0.095       0.5140         0.2245         0.1010   
1   M   0.350     0.265   0.090       0.2255         0.0995         0.0485   
2   F   0.530     0.420   0.135       0.6770         0.2565         0.1415   
3   M   0.440     0.365   0.125       0.5160         0.2155         0.1140   
4   I   0.330     0.255   0.080       0.2050         0.0895         0.0395   

   ShellWeight  Rings  
0        0.150     15  
1        0.070      7  
2        0.210      9  
3        0.155     10  
4        0.055      7

### 2.  Преобразуйте признак Sex в числовой

Значение F должно перейти в -1, I — в 0, M — в 1. Если вы используете Pandas, то подойдет следующий код: data[’Sex’] = data[’Sex’].map(lambda x: 1 if x == ’M’ else (-1 if x == ’F’ else 0))



In [6]:
df["Sex"] = df["Sex"].map(lambda x: 1 if x == "M" else (-1 if x == "F" else 0))
df.head()

Sex  Length  Diameter  Height  WholeWeight  ShuckedWeight  VisceraWeight  \
0    1   0.455     0.365   0.095       0.5140         0.2245         0.1010   
1    1   0.350     0.265   0.090       0.2255         0.0995         0.0485   
2   -1   0.530     0.420   0.135       0.6770         0.2565         0.1415   
3    1   0.440     0.365   0.125       0.5160         0.2155         0.1140   
4    0   0.330     0.255   0.080       0.2050         0.0895         0.0395   

   ShellWeight  Rings  
0        0.150     15  
1        0.070      7  
2        0.210      9  
3        0.155     10  
4        0.055      7

Sex  Length  Diameter  Height  WholeWeight  ShuckedWeight  VisceraWeight  \
0    0   0.455     0.365   0.095       0.5140         0.2245         0.1010   
1    0   0.350     0.265   0.090       0.2255         0.0995         0.0485   
2    0   0.530     0.420   0.135       0.6770         0.2565         0.1415   
3    0   0.440     0.365   0.125       0.5160         0.2155         0.1140   
4    0   0.330     0.255   0.080       0.2050         0.0895         0.0395   

   ShellWeight  Rings  
0        0.150     15  
1        0.070      7  
2        0.210      9  
3        0.155     10  
4        0.055      7

### 3.  Разделите содержимое файлов на признаки и целевую переменную.


В последнем столбце записана целевая переменная, в остальных — признаки.


In [8]:
X = df.drop("Rings", axis = 1) # признаки
y = df["Rings"] # целевая переменная

### 4.  Обучите случайный лес (sklearn.ensemble.RandomForestRegressor) с различным числом деревьев: от 1 до 50 (random_state=1).


Для каждого из вариантов оцените качество работы полученного леса на кросс-валидации по 5 блокам. Используйте параметры
"random_state=1"и "shuffle=True"при создании генератора кросс-валидации sklearn.cross_validation.KFold. В качестве меры качества воспользуйтесь коэффициентом детерминации (sklearn.metrics.r2_score).


### 5.  Определите, при каком минимальном количестве деревьев случайный лес показывает качество на кросс-валидации выше 0.52.


Это количество и будет ответом на задание.


###  6. Обратите внимание на изменение качества по мере роста числа деревьев. Ухудшается ли оно?
